React Native Tutorial (#2105)
mongodb/docs-realm: source/tutorial/react-native.txt

Overview
--------

> You can use the Realm JavaScript SDK and the official Realm React hooks

Reviewer: ```suggestion
You can use the Realm JavaScript SDK and the official Realm React
```

I know we call it Realm Hooks sometimes internally, but iirc we're supposed to just refer to it as either the "Realm React" or "@realm/react" library.

library to build a mobile application with React Native. This tutorial
walks you through how to build your own app that uses flexible sync.

> For this tutorial, we'll start with a pre-built template application to

Reviewer: Suggestion (not required for LGTM): I think the other SDKs had a slightly more descriptive variation of a sentence that explains the tutorial:
```
This tutorial adds functionality to the Template App. You will add a new Priority field to the existing Item model and update the Flexible Sync Subscription to only show items within a range of priorities.
```

   tutorial, check out the :ref:`React Native Quick Start
   <react-native-client-quick-start>`. It includes copyable code
   examples and the essential information that you need to set up a
   React Native app with {+sync+}.

Before You Begin
----------------

You must set up your local environment for React Native development
before you can begin this tutorial. For detailed instructions, see
`Setting up the development environment
<https://reactnative.dev/docs/environment-setup>`_ in the React Native
docs.

Start with the Template App
---------------------------

To start, you'll create a new app on your computer and get familiar with
it. The app will use a pre-built template that includes a working React
Native app and its corresponding Atlas App configuration files.

The template app is a basic "to-do list" application. Users can input a
list of tasks that they need to do and then check each to-do off when
they finish it. Some basic functionality is already set up for you:

- Users can create email/password accounts and log in and out of the app.

- Users can create, read, update, and delete ``Item`` objects that

>   represent individual to-do tasks.

Reviewer: ```suggestion
  represent individual items to do.
```

- A data source linked to your Atlas Cluster.

- A schema for the ``todo.Item`` collection that matches the ``Item``
  class in the React Native app.

- An authentication provider that lets users register for and log in to
  your app with an email and password.

- A flexible sync configuration with a single session role that allows
  users to read and write only their own data.

React Native App
~~~~~~~~~~~~~~~~

The React Native app is a fully configured mobile client that can run on
iOS and Android devices. It is located in the
``frontend/react-native.todo.flex`` directory.

The app uses the official :npm:`@realm/react` library. The library
includes React hooks and components that let you connect to your Atlas
App, read and write data, and automatically sync changes from other
devices.

The app contains some configuration files and directories, but you can
ignore those unless you want to customize the app. For this tutorial,
you'll want to be familiar with the React components in the ``src/``
directory:

.. list-table::
   :header-rows: 1
   :widths: 20 80

   * - File Name
       
     - Description
     
   * - ``ItemSchema.js``
       

>      - This is where we define the ``Item`` class, including its object

Reviewer: Suggestion/Not required for LGTM: For this table, I don't think it's necessary to start each description sentence with "This is", it gets repetitive and doesn't add value to the sentence. 

For example, the first description (ItemSchema.js), the following might be a bit more concise and avoid the fragment 'this is':

```
The Item class, including its object schema. We import the class in RealmContext.js
to include it in the app's Realm schema.
```

       schema. We import the class in ``RealmContext.js`` to include it
       in the app's Realm schema.
     
   * - ``RealmContext.js``
       
     - This is where we create the Realm hooks and providers and then
       export them to use throughout the app.
     
   * - ``App.js``
       
     - This is the root component of the app, which we wrap with the
       Realm hook providers from ``RealmContext.js``.
     
   * - ``WelcomeView.js``
       
     - This is the user registration and login form that users see when
       they first open the app.
     
   * - ``ItemListView.js``
       
     - This is the main "to-do list" app that users interact with after
       they log in. It queries for ``Item`` objects from Realm and
       displays them in a list.
       
       It also includes the code to create new ``Item`` objects and
       store them in Realm.
     
   * - ``CreateToDoPrompt.js``
       
     - This is a UI form that lets us enter data for new ``Item``
       objects. The code that actually creates new objects is in
       ``ItemListView.js``.
     
   * - ``LogoutButton.js``
       

>      - This is a re-usable button that logs out an authenticated user.

Reviewer: Needed fix:
```suggestion
     - This is a reusable button that logs out an authenticated user.
```

Add a Priority Level Field
--------------------------

Now that you're more familiar with what's already provided in the
template app, let's write some code to implement a new feature.

For this tutorial, we'll add a new "priority" property to each of the
``Item`` objects. This will let us organize to-dos by how important they
are and allow us to focus only on the most important ones.

.. procedure::

   .. step:: Define the Priority Levels
      

>       We want to allow a small number of named priority levels and we

Reviewer: ```suggestion
      We want to allow a small number of named priority levels, and we
```

>       want to easily be able sort the levels. In TypeScript, you could

Reviewer: Question: Do we really need to describe what to do in TS if the template app doesn't use typescript?

Author: I mainly wanted a handy reference to say "doing it this way gives you the TS enum behavior" - I'll update to focus less on TS while still keeping the pointer.

      define an `enum <https://www.typescriptlang.org/docs/handbook/enums.html>`_
      to model this:

      .. code-block:: javascript
         
         // Priority.High === 1
         // Priority[Priority.High] === "High"
         enum Priority {
           Severe = 0,
           High = 1,
           Medium = 2,
           Low = 3,
         }
      
      In regular JavaScript, there is no ``enum`` keyword. Instead we'll
      use a helper function to create something equivalent.

      Add the following code directly under the import statements in
      ``src/ItemSchema.js``:
      
      .. code-block:: javascript
         :caption: ItemSchema.js
         
         function createEnum(arr) {
           arr.forEach((p, i) => arr[p] = i);
           return arr;
         }
         
         // Priority.High === 1
         // Priority[Priority.High] === "High"

>          const Priority = createEnum([

Reviewer: Needed fix: You need to export 'Priority' in order to import it into `CreateToDoPrompt.js` as shown in the step [3.Add a Priority Picker](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/rn-tut/tutorial/react-native/#add-a-priority-picker)

>          const Priority = createEnum([

Reviewer: ```suggestion
         export const Priority = createEnum([
```

Reviewer: Needed fix: You need to export 'Priority' in order to import it into `CreateToDoPrompt.js` as shown in the step [3.Add a Priority Picker](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/rn-tut/tutorial/react-native/#add-a-priority-picker)

           "Severe",
           "High",
           "Medium",
           "Low",
         ])
      
      .. note::

         The priority levels in the enum are ordered from most important
         to least. The corresponding index value for each level
         increases from the most important, ``Priority[0]``, to the

>          least, ``Priority[3]``. This means that a "higher" priority

Reviewer: ```suggestion
         least important, ``Priority[3]``. This means that a "higher" priority
```

         level (i.e. a more important one) has a lower index value.
   
   .. step:: Update the To-Do Item Data Model

      Now we have an ``enum`` that defines the possible values of the
      ``priority`` field. However, we still have to define the
      ``priority`` field in the ``Item`` class.

      Add the following lines to your code in ``src/ItemSchema.js`` to
      add ``priority`` to the model:
      
      .. code-block:: javascript
         :caption: ItemSchema.js
         :emphasize-lines: 6, 11, 21
         
         export class Item {
           constructor({
             _id = new BSON.ObjectId(),
             isComplete = false,
             owner_id,

>              priority = Priority.High, // Default to High priority if none is specified

Reviewer: ```suggestion
             priority = Priority["High"], // Default to High priority if none is specified
```

If it's Priority.High you will get the error: undefined is not an object 
![Uploading Screen Shot 2022-08-23 at 12.49.40 PM.png…]()

           }) {
             this.isComplete = isComplete;
             this._id = _id;
             this.owner_id = owner_id
             this.priority = priority
           }
           
           static schema = {
             name: 'Item',
             properties: {
               _id: 'objectId',
               isComplete: {type: 'bool', default: false},
               summary: 'string',
               owner_id: 'string',

>                priority: 'int' // Store the index value of the Priority enum rather than the name

Reviewer: Should this be?: 

```
priority: {type: 'int', default: Priority.High},
```

             },
             primaryKey: '_id',
           };
         }
      
      .. tip::

>          At this point your React Native ``Item`` model and its

Reviewer: The other tutorials have a similar statement about why this doesn't break sync, do you just want to remove this paragraph and include that note, it reads:

```
Why Didn't This Break Sync?
Adding a property to a Realm object is not a breaking change and therefore does not 
require a client reset. The template app has Development Mode enabled, so changes
to the client Realm object are reflected in the server-side schema. For more information,
 see Enable or Disable Development Mode and Update Your Schema.

```

Author: Good point, updated

         corresponding schema in your Atlas App no longer agree. That's
         okay for now.

         The template app uses development mode by default. In this
         mode, any data model changes that your client app syncs will
         automatically update the backend schema. The next time you run
         the React Native app, the schemas will sync.

   .. step:: Add a Priority Picker

>       Your apps data model now includes a ``priority`` for each ``Item``

Reviewer: ```suggestion
      Your app's data model now includes a ``priority`` for each ``Item``
```

      object. Let's update the app UI so that you can choose a priority
      value when you add a new to-do to the list.
      
      First, we'll install an external library to implement the priority
      picker component. Run the following in your terminal inside of
      your project root:
      
      .. code-block:: shell
         

>          npm install @react-native-picker/picker

Reviewer: Needed fix: The user needs to pod install after installing @react-native-picker/picker, otherwise they'll get the following error:

![Uploading picker-not-found.png…]()

Author: Oops good catch! I forgot I did this 😅

      
      Next, let's update the new to-do creation prompt component to use
      the picker.
      
      Add the following imports to the top of ``src/CreateToDoPrompt.js``:

      .. code-block:: javascript
         :caption: CreateToDoPrompt.js
         
         import {Picker} from '@react-native-picker/picker';
         import {Priority} from './ItemSchema';
   
      Then, modify the ``CreateToDoPrompt`` component:
      
      - keep track of ``priority`` in a state hook
      - connect the state to the ``Picker`` component that you imported
      - pass ``priority`` to the ``onSubmit()`` handler

      .. code-block:: javascript
         :caption: src/CreateToDoPrompt.js
         :emphasize-lines: 4, 12-19, 23
         
         export function CreateToDoPrompt(props) {
           const {onSubmit} = props;
           const [summary, setSummary] = useState(null);

>            const [priority, setPriority] = useState(Priority.High);

Reviewer: ```suggestion
           const [priority, setPriority] = useState(Priority["High"]);
```
Priority.High will lead to an undefined is not an object error. See: https://github.com/mongodb/docs-realm/pull/2105/files#r952881610

           
           return (
             <View style={styles.modalWrapper}>
               <Text h4 style={styles.addItemTitle}>
                 Add Item
               </Text>
               <Input placeholder="What do you want to do?" onChangeText={setSummary} />
               <Picker

>                  style={{width: '80%'}}

Reviewer: Suggestion/not required for LGTM: This maybe a personal pet-peave of mine, but can we change the inline-styling to a styles.pickerWidth or something?

Author: I thought about doing this but it's an extra step for something that's not the point of the tutorial and is still a one-liner here. Not sure it justifies the extra step but if you think it's worth it we can change.

Reviewer: Good decision then :)

                 selectedValue={priority}
                 onValueChange={value => setPriority(value)}>
                 {Priority.map(p => (
                   <Picker.Item key={p} label={p} value={Priority[p]} />

>                  ))}

Reviewer: Suggestion/not required for LGTM (but strongly suggested): Can we choose something more descriptive than 'p', maybe:
```
                 {Priority.map((priorityValue, i) => (
                   <Picker.Item key={`picker-${priorityValue}-${i}`} label={priorityValue} value={Priority[priorityValue]} />
                 ))}
```

Note: I chose "`picker-${priorityValue}-${i}`" above because keys should be unique

Author: +1 for a more descriptive name. Keys only need to be unique within the map and the enum guarantees that each priority value is unique in the enum, so in this case just `priorityValue` will work.

your app needs. Let's add a feature to demonstrate how.
  
For this tutorial, we'll add a button that lets us toggle between two
"modes": one where the app syncs *all* to-do items and another where it
only syncs "important" ones with a ``priority`` of High or Severe.

.. procedure::

   .. step:: Add a Mode Toggle to the UI

      First, add a ``useState()`` hook to the ``ItemListView`` component
      to keep track of the current mode:

      .. code-block:: javascript
         :caption: ItemListView.js

         const [showImportantOnly, setShowImportantOnly] = useState(false);

      Then, add a new button that toggles the mode to the bottom of the
      to-do list:

      .. code-block:: javascript
         :caption: ItemListView.js
         :emphasize-lines: 6-13

         {items.map(item => (
           <ListItem key={`${item._id}`} bottomDivider topDivider>
             {/* ... */}
           </ListItem>
         ))}
         <Button
           title={showImportantOnly ? 'Show All' : 'Show Important Only'}
           buttonStyle={{
             ...styles.addToDoButton,
             backgroundColor: showImportantOnly ? '#00A35C' : '#FFC010',
           }}

>            onPress={() => setShowImportantOnly(b => !b)}

Reviewer: Suggestion/not required for lgtm (but strongly suggest): Can you change it from 'b' to something more descriptive, maybe: ``` onPress={() => setShowImportantOnly(buttonType => !buttonType)}```

         />

   .. step:: Update the Sync Subscription

      At this point, the app can switch modes in the UI but we haven't
      done anything else so the modes are functionally identical. Let's
      update the sync subscription to only sync data relevant to the
      current mode.

      In the first ``useEffect`` of the ``ItemListView`` component,
      add code that checks the current mode and appends an additional
      ``priority`` filter to the query if "important only" mode is
      active:

      .. code-block:: javascript
         :caption: ItemListView.js
         :emphasize-lines: 9-12, 18

         useEffect(() => {
           // initialize the subscriptions
           const updateSubscriptions = async () => {
             await realm.subscriptions.update(mutableSubs => {
               // subscribe to all of the logged in user's to-do items
               let ownItems = realm
                 .objects("Item")
                 .filtered(`owner_id == "${user.id}"`);
               if(showImportantOnly) {

>                  // More important Items have lower priority values

Reviewer: ```suggestion
                 // Filter 'Severe' or 'High' priority level items
```
The sentence "More important Items have lower priority values" seems very clunky and is hard to understand (especially if the reader skimmed the explanation that priority value is inversely correlated to importance)

                 ownItems = ownItems.filtered(`priority <= 1`);
               }
               // use the same name as the initial subscription to update it
               mutableSubs.add(ownItems, {name: 'ownItems'});
             });
           };
           updateSubscriptions();
         }, [realm, user, showImportantOnly]);

      .. important::

         Don't forget to add ``showImportantOnly`` to the list of
         dependencies in the second argument of ``useEffect``.

   .. step:: Test the App

      Your app is now set up to modify its sync subscription based on
      the current mode. However, if you run the app and switch to

>       "important only" mode, you'll see that the list doesn't change and

Reviewer: ```suggestion
      "important only" mode, you'll see that the list doesn't change, and
```

      
      We're getting a warning because a sync client can only filter a
      subscription based on a given field if that field is specifically
      listed as a "queryable field" in the Device Sync configuration of
      your Atlas App.

      To get everything working, we need to add ``priority`` as a
      queryable field in the Atlas App's sync configuration.
      
      First, find the template app's ``backend`` directory and modify
      ``sync/config.json``:

      .. code-block:: json
         :emphasize-lines: 6

         {
             ...,
             "queryable_fields_names": [
                 "_id",
                 "owner_id",
                 "priority"
             ]
         }
      
      Then, run the following terminal command from within the
      ``backend`` directory to deploy the updated sync configuration:

      .. code-block:: bash

         npx mongodb-realm-cli push

   .. step:: Run the Completed App
      
      Once your updated sync configuration is deployed, your app is
      complete and fully functional. Congratulations!
      
      Rebuild and run the app to make sure everything works. You should

>       be able create, complete, and delete to-do items as well as switch

Reviewer: ```suggestion
      be able to create, complete, and delete to-do items as well as switch
```

  clients with {+sync+}.

Use the following steps to get the template app up and running on your
computer:

.. procedure::

   .. step:: Create the Template App
      
      Run the following command in a terminal to create a new folder
      with your templated configuration files and React Native app:
      
      .. code-block:: shell
         
         npx mongodb-realm-cli apps create \
           --template react-native.todo.flex \
           --name realm-sync-tutorial
      
      .. tip::
         
         The command creates a new folder in your current directory with
         the same name as the value of the ``--name`` flag.
   
   .. step:: Install Dependencies

      In your terminal, run the following commands to navigate to the
      react native app and install its dependencies:

      .. code-block:: shell

         cd realm-sync-tutorial/frontend/react-native.todo.flex
         npm install
      
      To build and run the app on an iOS device or simulator, install
      the additional iOS dependencies with `CocoaPods
      <https://guides.cocoapods.org/using/getting-started.html>`__.

      .. code-block:: shell

> 

Author: Neat! Good suggestion thanks

      or
      .. tip::

      .. code-block:: shell
         Make sure to switch back to the frontend project's root
         directory after you install the pods.

         npx react-native run-android
   .. step:: Build the App

      After the emulator is started and the bits have been deployed, you should see
      the app is "running" and one or more errors are displayed:
      At this point, you should have a fully functional React Native app
      that can run on both iOS and Android.

      .. figure:: /images/rn-app-initial-state.png
        :alt: The initial state of React Native app
        :width: 250px
      To make sure that everything works, build the app and run it in an
      iOS simulator or Android emulator:

      .. important:: CompileC Error
      .. tabs::

         If, when compiling the iOS app, you get a CompileC build error that involves
         "Flipper," you will need to edit ``/ios/Podfile``. Look for the section
         that enables Flipper and comment out those lines:
         .. tab:: iOS
            :tabid: ios
            
            .. code-block:: shell

         .. code-block:: text
               npx react-native run-ios

>             # Enables Flipper.

Author: Updated RN!

>                npx react-native run-android

Author: Good catch, added

>       - Add a few to-do items to the list.

Author: Ah yeah I actually already installed the package just never added the import 😅

> ``backend`` directory. It has a unique ``app_id`` value in

Author: Yeah this is a side-effect of how we organize the template apps so not a simple change unfortunately. I can bring it up with the team.

template app, let's write some code to implement a new feature.

For this tutorial, we'll add a new "priority" property to each of the
``Item`` objects. This will let us organize to-dos by how important they
are and allow us to focus only on the most important ones.

.. procedure::

   .. step:: Define the Priority Levels
      
      We want to allow a small number of named priority levels, and we
      want to easily be able sort the levels. To do this, we'll use a
      helper function to define an "enum" object that maps a set of
      ordered level names to and from an integer that represents their
      priority.
      
      Add the following code directly under the import statements in
      ``src/ItemSchema.js``:
      
      .. code-block:: javascript
         :caption: ItemSchema.js
         
         function createEnum(arr) {
           arr.forEach((p, i) => arr[p] = i);
           return arr;
         }
         
         // Priority.High === 1
         // Priority[Priority.High] === "High"
         export const Priority = createEnum([
           "Severe",
           "High",
           "Medium",
           "Low",

>          ])

Reviewer 2: The `createEnum` function is a bit overkill.  I would just define this as an object in descending order:
```
export const Priority = {
  Low: 0,
  Medium: 1,
  High: 2,
  Severe: 3
}
```
Then one can reference the priority as `Priority.Medium` which makes the code clearer when doing logic comparisons or assignments:
```
if( item.priority > Priority.High ) 
//...

realm.write(() => {
  item.priority = Priority.Low
})
```

Author: The code samples you have will work with the enum too. I chose to use the enum approach because it gives a two-way mapping between the level name and value. That way I can use the `Priority.High` syntax and _also_ map through the Priority enum to get the name and value together in the Picker component.

Reviewer 2: Ah, sure.  Makes sense that the Picker component needs an array instead of an object.

      
      The priority levels in the enum are ordered from most important
      to least. The corresponding index value for each level increases
      from the most important, ``Priority[0]``, to the least important,
      ``Priority[3]``. This means that a "higher" priority level (i.e. a
      more important one) has a lower index value.

      .. tip::
      

>          In a TypeScript app, you could define an `enum

Reviewer 2: We should also have a TypeScript version of this template and tutorial.  When doing so, one should consult us about how to correctly do class based models (there are changes coming to this in v11)

Author: Yeah agreed - I think we've held off on going full force into TS until the models stuff is fully baked but we're all about it.

         corresponding schema in your Atlas App no longer agree. That's
         okay!

         Adding a property to a Realm object is not a breaking change
         and therefore does not require a :ref:`client reset
         <client-resets>`. The template app has Development Mode
         enabled, so changes to the client Realm object are reflected in
         the server-side schema. For more information, see
         :ref:`<concept-development-mode>` and
         :ref:`<additive-changes-synced-schema>`.

   .. step:: Add a Priority Picker

      Your app's data model now includes a ``priority`` for each ``Item``
      object. Let's update the app UI so that you can choose a priority
      value when you add a new to-do to the list.
      
      First, we'll install an external library to implement the priority
      picker component. Run the following in your terminal inside of
      your project root:
      
      .. code-block:: shell
         
         npm install @react-native-picker/picker
      
      After you've installed the package, make sure to link the
      associated Cocoapods:

      .. code-block:: shell

         cd ios

>          pod install --repo-update

Reviewer 2: This is only for `ios`.  It should be clear that they will have to rerun `npm run ios` or `npm run android` in order for the picker ot be useable.  They also might have to restart their realm bundler

Overview
--------

You can use the Realm JavaScript SDK, and the official Realm React
library to build a mobile application with React Native. This tutorial
walks you through how to build your own app that uses flexible sync.

For this tutorial, we'll start with a pre-built template application to
see how everything fits together. Then, you will add a new Priority
field to the existing Item model and update the Flexible Sync
Subscription to only show items within a range of priorities.

This tutorial should take around 30 minutes.

.. note:: Check Out the Quick Start
   
   If you prefer to explore on your own rather than follow a guided
   tutorial, check out the :ref:`React Native Quick Start
   <react-native-client-quick-start>`. It includes copyable code
   examples and the essential information that you need to set up a
   React Native app with {+sync+}.

Before You Begin
----------------

You must set up your local environment for React Native development
before you can begin this tutorial. For detailed instructions, see
`Setting up the development environment
<https://reactnative.dev/docs/environment-setup>`_ in the React Native

> docs.

Reviewer 2: We should add a note here about Node 17.  And perhaps guide the user to [`nvm`](https://github.com/nvm-sh/nvm) if they need to downgrade Node.

Author: Node 17 is supported now since we're on the latest version of RN

         okay!

         Adding a property to a Realm object is not a breaking change
         and therefore does not require a :ref:`client reset
         <client-resets>`. The template app has Development Mode
         enabled, so changes to the client Realm object are reflected in
         the server-side schema. For more information, see
         :ref:`<concept-development-mode>` and
         :ref:`<additive-changes-synced-schema>`.

   .. step:: Add a Priority Picker

      Your app's data model now includes a ``priority`` for each ``Item``
      object. Let's update the app UI so that you can choose a priority
      value when you add a new to-do to the list.
      
      First, we'll install an external library to implement the priority
      picker component. Run the following in your terminal inside of
      your project root:
      
      .. code-block:: shell
         
         npm install @react-native-picker/picker
      
      If you're building for iOS, make sure to link the associated
      Cocoapods after you've installed the package:

      .. code-block:: shell

         npx pod-install
      
      .. tip::
         

>          You may need to restart your project bundler after installing.

Reviewer 2: They will also need to rebuild the app with `npm run ios` and `npm run android`.  I would recommend running these commands first, this way they can see the changes as they update their code and there won't be any errors.
